source/validate.cpp

```diff
@@ -518,6 +436,109 @@ void DebugInstructionPass(ValidationState_t& _,
   }
 }
 
+// Allows for piping of multiple functions which return spv_result_t
+spv_result_t operator||(spv_result_t lhs, spv_result_t rhs) {
+  if (lhs) return lhs;
+  if (rhs) return rhs;
+  return SPV_SUCCESS;
+}
+
+// TODO(umar): Check MemoryModel is in module
+// TODO(umar): Check OpVariable storage class is not function in module section
+// TODO(umar): Make sure function declarations appear before function
+// definitions
+// TODO(umar): Better error messages
+// NOTE: This function does not handle CFG related validation
+// Performs logical layout validation. See Section 2.4
+spv_result_t ModuleLayoutPass(ValidationState_t& _,
+                              const spv_parsed_instruction_t* inst) {
+  if (_.is_enabled(SPV_VALIDATE_LAYOUT_BIT)) {
+    SpvOp opcode = inst->opcode;
+
+    if (libspirv::ModuleLayoutSection::kModule == _.getLayoutStage()) {
+      // Module scoped instructions are processed by determining if the opcode
+      // is part of the current stage. If it is not then the next stage is
+      // checked.
+      while (_.isOpcodeInCurrentLayoutStage(opcode) == false) {
+        // TODO(umar): Check if the MemoryModel instruction has executed
+        _.progressToNextLayoutStageOrder();
+        if (_.getLayoutStage() == libspirv::ModuleLayoutSection::kFunction) {
+          // All module stages have been processed. Recursivly call
+          // ModuleLayoutPass
+          // to process the next section of the module
+          return ModuleLayoutPass(_, inst);
+        }
+      }
+    } else {
+      // Validate the function layout.
+      switch (opcode) {
```

Comment:
OpUndef is missing.  Should go in the function section.  It's not a "constant instruction", despite being listed in the constant-creation instructions section of the spec.